nvme_gw_persistence.py

```diff
@@ -0,0 +1,443 @@
+        pass
+
+    @abstractmethod
+    def set_allow_all(self, subsystem_nqn: str, val: str):
+        pass
+
+    @abstractmethod
+    def delete_allow_all(self, subsystem_nqn: str):
+        pass
+
+    @abstractmethod
+    def set_transport(self, trtype: str, val: str):
+        pass
+
+    @abstractmethod
+    def delete_transport(self, trtype: str):
+        pass
+
+    @abstractmethod
+    def delete_config(self):
+        pass
+
+    @abstractmethod
+    def restore(self):
+        pass
+
+
+class OmapPersistentConfig(PersistentConfig):
+    """Persists NVMeoF target configuration to an OMAP object.
+
+    Handles reads/writes of persistent NVMeoF target configuration data in 
+    key/value format within an OMAP object.
+
+    Class attributes:
+        X_KEY: OMAP key name for "X"
+        X_PREFIX: OMAP key prefix for key of type "X"
+
+    Instance attributes:
+        version: Local gateway NVMeoF target configuration version
+        nvme_config: Basic gateway parameters
+        logger: Logger instance to track OMAP access events
+        spdk_rpc: Module methods for SPDK
+        spdk_rpc_client: Client of SPDK RPC server
+        omap_name: OMAP object name
+        ioctx: I/O context which allows OMAP access
+    """
+
+    OMAP_VERSION_KEY = "omap_version"
+    BDEV_PREFIX = "bdev_"
+    NAMESPACE_PREFIX = "namespace_"
+    SUBSYSTEM_PREFIX = "subsystem_"
+    HOST_PREFIX = "subsystem_host_"
+    ALLOW_ALL_PREFIX = "allow_all_"
+    TRANSPORT_PREFIX = "transport_"
+    LISTENER_PREFIX = "listener_"
+
+    def __init__(self, nvme_config):
+        self.version = 1
+        self.nvme_config = nvme_config
+        self.logger = nvme_config.logger
+
+        gateway_group = self.nvme_config.get("config", "gateway_group")
+        self.omap_name = gateway_group + "_config"
+
+        ceph_pool = self.nvme_config.get("ceph", "ceph_pool")
+        ceph_conf = self.nvme_config.get("ceph", "ceph_config_file")
+        conn = rados.Rados(conffile=ceph_conf)
+        conn.connect()
+        self.ioctx = conn.open_ioctx(ceph_pool)
+
+        try:
+            # Create a new gateway persistance OMAP object
+            with rados.WriteOpCtx() as write_op:
+                # Set exclusive parameter to fail write_op if object exists
+                write_op.new(rados.LIBRADOS_CREATE_EXCLUSIVE)
+                self.ioctx.set_omap(write_op, (self.OMAP_VERSION_KEY,),
+                                    (str(self.version),))
+                self.ioctx.operate_write_op(write_op, self.omap_name)
+                self.logger.info(
+                    f"First gateway: created object {self.omap_name}")
+        except rados.ObjectExists:
+            self.logger.info(f"{self.omap_name} already exists.")
+
+        omap_version = self._read_key(self.OMAP_VERSION_KEY)
+        self.logger.info(
+            f"omap_version: {omap_version}, local version: {self.version}.")
```

Comment:
Why do we need to read `omap_version` here? It seems like is not used and we read it again when restoring the config.